kubespawner/spawner.py

```diff
@@ -1757,6 +1783,82 @@ def start(self):
 
     _last_event = None
 
+    @gen.coroutine
+    def _make_create_pod_request(self, pod, request_timeout):
+        """
+        Make an HTTP request to create the given pod
+
+        Designed to be used with exponential_backoff, so returns
+        True / False on success / failure
+        """
+        try:
+            self.log.info(f"Attempting to create pod {pod.metadata.name}, with timeout {request_timeout}")
+            # Use tornado's timeout, _request_timeout seems unreliable?
+            yield gen.with_timeout(timedelta(seconds=request_timeout), self.asynchronize(
+                self.api.create_namespaced_pod,
+                self.namespace,
+                pod,
+            ))
+            return True
+        except gen.TimeoutError:
+            # Just try again
+            return False
+        except ApiException as e:
+            pod_name = pod.metadata.name
+            if e.status != 409:
+                # We only want to handle 409 conflict errors
+                self.log.exception("Failed for %s", pod.to_str())
+                raise
+            self.log.info(f'Found existing pod {pod_name}, attempting to kill')
+            # TODO: this should show up in events
+            yield self.stop(True)
+
+            self.log.info(f'Killed pod {pod_name}, will try starting singleuser pod again')
+            # We tell exponential_backoff to retry
+            return False
+
+    @gen.coroutine
+    def _make_create_pvc_request(self, pvc, request_timeout):
+        # Try and create the pvc. If it succeeds we are good. If
+        # returns a 409 indicating it already exists we are good. If
+        # it returns a 403, indicating potential quota issue we need
+        # to see if pvc already exists before we decide to raise the
+        # error for quota being exceeded. This is because quota is
+        # checked before determining if the PVC needed to be
+        # created.
+        pvc_name = pvc.metadata.name
+        try:
+            self.log.info(f"Attempting to create pvc {pvc.metadata.name}, with timeout {request_timeout}")
+            yield gen.with_timeout(timedelta(seconds=request_timeout), self.asynchronize(
+                self.api.create_namespaced_persistent_volume_claim,
+                namespace=self.namespace,
+                body=pvc
+            ))
+            return True
+        except gen.TimeoutError:
+            # Just try again
+            return False
```

Comment:
Its a bit of pr-review-creep perhaps, but I feel like I want to mention that there are various others situations that will make sense to retry that will be listed as ApiException errors.

Here are those recommended to retry according to [this kubernetes community documentation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#http-status-codes).

429 StatusTooManyRequests
500 StatusInternalServerError
503 StatusServiceUnavailable
504 StatusServerTimeout

Reply:
That's a great comment! I think it's okay to address this here if desired by the author, but also fine to improve the retry conditions in a later PR, since this one doesn't change the PVC retry logic, just reorganizes it.

Reply:
I opened https://github.com/jupyterhub/kubespawner/issues/436 to represent this matter.